File: engine/app/api/routes/portfolio.py
from datetime import date

from fastapi import APIRouter, Depends

from ...core.security import require_internal_key
from ...schemas.options import GreeksData, PortfolioGreeksRequest, PortfolioGreeksResponse, PositionGreeks
from ...services.greeks import calculate_greeks
from ...services.kite_client import KiteClient as UpstoxClient

router = APIRouter(prefix="/portfolio", tags=["portfolio"])


@router.post("/greeks", response_model=PortfolioGreeksResponse, dependencies=[Depends(require_internal_key)])
async def calculate_portfolio_greeks(body: PortfolioGreeksRequest):
    """
    Calculate Greeks for each position and aggregate across the portfolio.
    Accepts current spot prices from the API layer (avoids multiple Upstox calls).
    """
    upstox = UpstoxClient()
    r = 0.065

    position_results: list[PositionGreeks] = []
    agg = {"delta": 0.0, "gamma": 0.0, "theta": 0.0, "vega": 0.0, "rho": 0.0}
    total_pnl = 0.0

    for pos in body.positions:
        symbol = pos.get("symbol", "NIFTY").upper()
        strike = float(pos.get("strike", 0))
        expiry = pos.get("expiry", date.today().isoformat())
        option_type = pos.get("option_type", "CE").upper()
        lots = int(pos.get("lots", 1))
        lot_size = int(pos.get("lot_size", 50))
        avg_price = float(pos.get("avg_price", 0))
        position_type = pos.get("position_type", "LONG").upper()
        position_id = pos.get("id", "")

        t = max((date.fromisoformat(expiry) - date.today()).days, 1) / 365
        spot = body.spot_prices.get(symbol) or await upstox.get_spot_price(symbol)

        # Fetch current LTP from Upstox or use IV=16% as fallback
        current_option_chain = await upstox.get_options_chain(symbol, expiry)
        strike_data = next((s for s in current_option_chain.strikes if s.strike_price == strike), None)
        if strike_data:
            od = strike_data.ce if option_type == "CE" else strike_data.pe
            current_price = od.ltp if od and od.ltp else 0.0
            iv = (od.iv or 16.0) / 100 if od else 0.16
        else:
            iv = 0.16
            current_price = 0.0

        g = calculate_greeks(spot, strike, t, iv, r, option_type)
        direction = 1 if position_type == "LONG" else -1
        qty = lots * lot_size * direction

        pnl = (current_price - avg_price) * abs(qty) * direction
        total_pnl += pnl

        for key in agg:
            agg[key] += g[key] * qty

        position_results.append(PositionGreeks(
            position_id=position_id,
            delta=round(g["delta"] * qty, 4),
            gamma=round(g["gamma"] * qty, 6),
            theta=round(g["theta"] * qty, 4),
            vega=round(g["vega"] * qty, 4),
            rho=round(g["rho"] * qty, 4),
            current_price=round(current_price, 2),
            pnl=round(pnl, 2),
        ))

    return PortfolioGreeksResponse(
        positions=position_results,
        aggregate=GreeksData(
            delta=round(agg["delta"], 4),
            gamma=round(agg["gamma"], 6),
            theta=round(agg["theta"], 4),
            vega=round(agg["vega"], 4),
            rho=round(agg["rho"], 4),
        ),
        total_pnl=round(total_pnl, 2),
    )

File: api/src/routes/portfolio.ts
import { Router, Response, NextFunction } from 'express'
import { z } from 'zod'
import axios from 'axios'
import { PrismaClient } from '@prisma/client'
import { AuthRequest, requireAuth } from '../middleware/requireAuth'
import { getValidToken } from '../services/kite.service'

const router = Router()
const prisma = new PrismaClient()

const ENGINE_URL = process.env.ENGINE_URL ?? 'http://localhost:8000'
const INTERNAL_KEY = process.env.INTERNAL_API_KEY ?? 'dev-internal-key'

const positionSchema = z.object({
  symbol: z.enum(['NIFTY', 'BANKNIFTY', 'FINNIFTY', 'MIDCPNIFTY']),
  strike: z.number().positive(),
  expiry: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
  optionType: z.enum(['CE', 'PE']),
  positionType: z.enum(['LONG', 'SHORT']),
  lots: z.number().int().min(1),
  lotSize: z.number().int().min(1).default(50),
  avgPrice: z.number().min(0),
  instrumentKey: z.string().optional(),
  notes: z.string().max(500).optional(),
})

// GET /api/v1/portfolio/positions
router.get('/positions', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const positions = await prisma.position.findMany({
      where: { userId: req.userId! },
      orderBy: { createdAt: 'desc' },
    })
    res.json(positions)
  } catch (err) {
    next(err)
  }
})

// POST /api/v1/portfolio/positions
router.post('/positions', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const body = positionSchema.parse(req.body)
    const position = await prisma.position.create({
      data: { ...body, userId: req.userId!, source: 'MANUAL' },
    })
    res.status(201).json(position)
  } catch (err) {
    next(err)
  }
})

// PUT /api/v1/portfolio/positions/:id
router.put('/positions/:id', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const posId = req.params['id'] as string
    const existing = await prisma.position.findFirst({
      where: { id: posId, userId: req.userId! },
    })
    if (!existing) return res.status(404).json({ message: 'Position not found' })

    const body = positionSchema.partial().parse(req.body)
    const updated = await prisma.position.update({ where: { id: posId }, data: body })
    res.json(updated)
  } catch (err) {
    next(err)
  }
})

// DELETE /api/v1/portfolio/positions/:id
router.delete('/positions/:id', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const posId = req.params['id'] as string
    const existing = await prisma.position.findFirst({
      where: { id: posId, userId: req.userId! },
    })
    if (!existing) return res.status(404).json({ message: 'Position not found' })
    await prisma.position.delete({ where: { id: posId } })
    res.json({ message: 'Position deleted' })
  } catch (err) {
    next(err)
  }
})

// POST /api/v1/portfolio/import-from-zerodha
router.post('/import-from-zerodha', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const token = await getValidToken(req.userId!)
    if (!token) {
      return res.status(400).json({
        message: 'Zerodha account not connected or session expired (tokens reset at 6 AM IST). Reconnect via /api/v1/kite/auth',
      })
    }

    const apiKey = process.env.KITE_API_KEY ?? ''
    const { data } = await axios.get('https://api.kite.trade/portfolio/positions', {
      headers: {
        'X-Kite-Version': '3',
        Authorization: `token ${apiKey}:${token}`,
      },
    })

    // Kite returns {net: [...], day: [...]}; use net positions
    const kitePositions: any[] = data.data?.net ?? []
    const optionPositions = kitePositions.filter(
      (p: any) => (p.product === 'NRML' || p.product === 'MIS') && p.quantity !== 0 && p.exchange === 'NFO',
    )

    const LOT_SIZES: Record<string, number> = {
      NIFTY: 25, BANKNIFTY: 15, FINNIFTY: 40, MIDCPNIFTY: 50,
    }

    let imported = 0
    for (const p of optionPositions) {
      // Kite tradingsymbol: NIFTY25MAY2222000CE, BANKNIFTY25MAY2247000PE, etc.
      const ts: string = p.tradingsymbol ?? ''
      const symbolMatch = ts.match(/^(NIFTY|BANKNIFTY|FINNIFTY|MIDCPNIFTY)/)
      const symbol = symbolMatch?.[1] ?? 'NIFTY'
      const optionType: 'CE' | 'PE' = ts.endsWith('CE') ? 'CE' : 'PE'
      const strikeMatch = ts.match(/(\d+)(CE|PE)$/)
      const strike = strikeMatch ? parseFloat(strikeMatch[1]) : 0
      // Kite expiry is ISO date string: "2025-05-29"
      const expiry: string = p.expiry ?? new Date().toISOString().split('T')[0]
      const lotSize = LOT_SIZES[symbol] ?? 50
      const lots = Math.abs(Math.round(p.quantity / lotSize)) || 1
      const avgPrice = p.average_price ?? 0
      const positionType: 'LONG' | 'SHORT' = p.quantity > 0 ? 'LONG' : 'SHORT'

      if (!strike) continue

      const posId = `kite-${req.userId!}-${ts}`
      await prisma.position.upsert({
        where: { id: posId },
        create: {
          id: posId,
          userId: req.userId!,
          symbol,
          instrumentKey: ts,
          strike,
          expiry,
          optionType,
          positionType,
          lots,
          lotSize,
          avgPrice,
          source: 'UPSTOX_IMPORT', // kept as-is to avoid schema change; semantically means "broker import"
        },
        update: {
          lots,
          avgPrice,
          positionType,
          currentPrice: p.last_price ?? null,
        },
      })
      imported++
    }

    res.json({ imported, total: optionPositions.length })
  } catch (err) {
    next(err)
  }
})

// POST /api/v1/portfolio/greeks  — proxy to engine with current spot prices
router.post('/greeks', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const positions = await prisma.position.findMany({ where: { userId: req.userId! } })
    if (positions.length === 0) {
      return res.json({ positions: [], aggregate: { delta: 0, gamma: 0, theta: 0, vega: 0, rho: 0 }, total_pnl: 0 })
    }

    const { data } = await axios.post(
      `${ENGINE_URL}/api/v1/portfolio/greeks`,
      {
        positions: positions.map((p) => ({
          id: p.id,
          symbol: p.symbol,
          strike: p.strike,
          expiry: p.expiry,
          option_type: p.optionType,
          position_type: p.positionType,
          lots: p.lots,
          lot_size: p.lotSize,
          avg_price: p.avgPrice,
        })),
        spot_prices: req.body?.spot_prices ?? {},
      },
      { headers: { 'X-Internal-Key': INTERNAL_KEY } },
    )
    res.json(data)
  } catch (err) {
    next(err)
  }
})

export default router
